### PR TITLE
[IGA] Add Geometry types

### PR DESCRIPTION
**Pull request by oberbichler**
- `NodeCurveGeometry3D`
- `NodeSurfaceGeometry3D`

### CODE REVIEW

File: applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp]: please use `lower_case_names` for the file names

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: Please use the template for the core for the classes

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: Why not `GetNode` and `SetNode` ?

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: I recommend adding debug errors here (and at other places too):
`KRATOS_DEBUG_ERROR_IF(Index >= mNodes.size()) << "..." << std::endl;`

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: const

File: applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp]: If you typedef this then I recommend to name it `Array3` or `Array3D`, this is usually used

Otherwise it can be confused with the ublas/amatrix Vector

File: applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp]: just curious, what does the _a do?

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: This name is not very descriptive
Also is this a template?

File: applications/NurbsBrepApplication/custom_utilities/NodeSurfaceGeometry3D.h
Comment (philbucher) [inline applications/NurbsBrepApplication/custom_utilities/NodeSurfaceGeometry3D.h]: Usually the return type is in the same line as the function name etc

File: applications/NurbsBrepApplication/tests/NodeCurveGeometry3DTests.py
Comment (philbucher) [inline applications/NurbsBrepApplication/tests/NodeCurveGeometry3DTests.py]: is this a Kratos Node or sth else?

File: applications/NurbsBrepApplication/CMakeLists.txt
Comment (philbucher) [inline applications/NurbsBrepApplication/CMakeLists.txt]: this works without including ANurbs as a submodule?

**Review by philbucher — CHANGES_REQUESTED** [review]
Some rather general comments:
- Please add class & function descriptions, or at least for the classes
- :+1: for tests

File: applications/NurbsBrepApplication/CMakeLists.txt
Comment (oberbichler) [reply applications/NurbsBrepApplication/CMakeLists.txt]: At the moment it woks by specifing `ANURBS_ROOT` in your build script (the same as for Eigen). I need to wait for a final decision to integrate the code into Kratos.

File: applications/NurbsBrepApplication/tests/NodeCurveGeometry3DTests.py
Comment (oberbichler) [reply applications/NurbsBrepApplication/tests/NodeCurveGeometry3DTests.py]: A Kratos node

File: applications/NurbsBrepApplication/custom_utilities/NodeSurfaceGeometry3D.h
Comment (oberbichler) [reply applications/NurbsBrepApplication/custom_utilities/NodeSurfaceGeometry3D.h]: Mozilla-Style... important?

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (oberbichler) [reply applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: The idea was that `curve.PointAt(T=...)` evaluates the actual XYZ of the nodes on the curve at the given parameter `T`.

Then `ValueAt(Variable=..., T=...)` should evaluate some specific `Variable` at `T`.

Maybe `ValueOfVariableAt`?

File: applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp
Comment (oberbichler) [reply applications/NurbsBrepApplication/custom_python/add_custom_utilities_to_python.cpp]: Just a shortcut for `py::arg("named_argument")`. I think named arguments make code much more easier to read.

e.g. `curve.PointAt(T=0.5)`

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (oberbichler) [reply applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: 👍 Also on my TODO for the library...
Do you think it would be better to throw an error also in release mode?

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (oberbichler) [reply applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: The common style in C++ is `node/setNode`, in Java `getNode/setNode`. Anyway: I can use `Get...` if it is proposed by the Kratos-styleguide. In this class it might be better to use the style of the parent class.

File: applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h
Comment (oberbichler) [reply applications/NurbsBrepApplication/custom_utilities/NodeCurveGeometry3D.h]: Btw: How can I check inside python if Kratos was compiled in debug mode?

**Comment by pooyan-dadvand** [comment]
Why you had to specify the point type as node instead of using template like other geometries?

**Comment by oberbichler** [comment]
The NURBS geometry does not need the control point itself. So at the first moment the type is not important. But it needs to know how to get the weight of a control point (for the shape functions) and its location (for the evaluation). Of course the way 'how to get' this value depends on the kind of data structure that contains it.

The `ANurbs` provides a standard implementation of the geometries which stores a list of points (= control point location) and a list of weights. This has the advantage that if the weights are unset than just the simpler and faster B-Spline algorithm is used (might be useful for the trimming curves). These classes are templated so that they can work with every vector type as long as they provide access to the components via squared brackets (`vector[0] = x` ...).

The idea of using nodes as datatype for the control points was that the nodes are updated automatically during the analysis. That holds also for other values stored at the nodes. The IGA beam e.g. has an additionally DOF which defines the rotation around the middle axis. With the `ValueAt` function it is possible to get the value of this rotation on a geometry point according to the current value of the variables on the nodes. Then I can call e.g. a plotting function after each timestep to visualize the movement and rotation of my cross sections. The nodes are storing the location (X/Y/Z getter) and the weight (stored in a nodal variable introduced by the application) in a very specific way. So I need to override the functions for this case.

I am not familiar with the details of the other Kratos geometries. But if it is possible it would make sense to make the NURBS geometries work in a similar way.

**Comment by oberbichler** [comment]
@pooyan-dadvand At the moment I think that this is the only way to keep the data of the NURBS-Patchs without breaking the code inside the `Brep_application` branch completely.

**Comment by pooyan-dadvand** [comment]
OK, I see. We should keep it in mind to make it more standard over the time

**Comment by oberbichler** [comment]
Name of `NURBS_CONTROLPOINT_WEIGHT` will be checked in #2560

**Pull request merged**
